### PR TITLE
Remove flask dependency from validation

### PR DESCRIPTION
**Pull request by tbille**
# Refactoto S02E06 - The validation

~~This PR should be merged before: https://github.com/canonical-websites/snapcraft.io/pull/431~~

## Summary

- Remove last flask dependency from publisher API module
- Remove useless validation checks for the frontend. Only check if the macaroon needs to be refreshed in the headers. If yes raise an exception to handle it.

## QA

- `./run`
- http://0.0.0.0:8004/account/
- Should work as usual

Hard to QA because the macaroon is valid for 1 year...

### CODE REVIEW

File: modules/authentication.py [hunk 1]
```diff
@@ -169,11 +170,11 @@ def verify_response(
             session['macaroon_discharge']
```
Comment (nottrobin) [inline modules/authentication.py]: The docstring above needs updating to reflect the new behaviour of this function

File: modules/authentication.py [hunk 2]
```diff
@@ -169,11 +170,11 @@ def verify_response(
...
         )
 
-        return {
-            'status_code': 307,
-            'redirect': url_calling,
-            'reason': 'Macaroon discharge refreshed'
-        }
+        raise ApiVerifyResponseError(
```
Comment (nottrobin) [inline modules/authentication.py]: I don't think checking if the macaroon needs refreshing should happen in the same function that we check for other general errors. And I'm also not sure that a Exception is the best way to communicate this state, or things like URLs. I haven't understood this code well enough to provide concrete suggestions, but I think probably we should have a "check authorization" step somewhere which is where things like macaroon freshness should be checked.

File: modules/authentication.py [hunk 3]
```diff
@@ -185,24 +186,24 @@ def verify_response(
...
         # Not authorized content
         if response.status_code == 401 and not verified['allowed']:
-            return {
-                'status_code': 401,
-                'redirect': url_account,
-                'reason': 'Not authorized'
-            }
+            raise ApiVerifyResponseError(
+                'Not authorized',
+                401,
```
Comment (nottrobin) [inline modules/authentication.py]: Could you write this as `status_code=response.status_code`? Firstly, I think it's helpful to use named arguments if you can because it helps to clarify how the function might be using that argument. Secondly, I think we should make it clear that this API library is never simply choosing status_codes to pass back itself - whenever an ApiError has a status_code property it's because that status_code came directly from the API itself.

File: modules/authentication.py [hunk 4]
```diff
@@ -185,24 +186,24 @@ def verify_response(
...
+            )
 
         # The package doesn't exist
         if verified['account'] is not None and response.status_code == 404:
-            return {
-                'status_code': 404,
-                'redirect': url_account,
-                'reason': 'Not found'
-            }
+            raise ApiVerifyResponseError(
```
Comment (nottrobin) [inline modules/authentication.py]: Actually all we're basically doing in this function, if you remove the macaroon refresh stuff, is pass request errors back to the client. Maybe it's not even worth wrapping these - we should simply do `response.raise_for_status` and let the client catch them? Or do you think there's an argument for wrapping it in our own error type?

If we do wrap it in our own error type, couldn't it be generic? E.g.:

if response.status_code >= 400:
    raise ApiResponseError(
        "The API returned a bad response",
        response=response
    )

File: modules/publisher/api.py [hunk 5]
```diff
@@ -1,11 +1,11 @@
-import flask
 import modules.authentication as authentication
 import modules.cache as cache
 import os
 from modules.exceptions import (
     ApiResponseDecodeError,
     ApiResponseError,
     ApiResponseErrorList,
+    MacaroonNeedsRefresh
```
Comment (WillMoggridge) [inline modules/publisher/api.py]: Minor nitpick on name, I think "requires" is better than "needs".
I'm also wondering if it sounds better as `MacaroonRefreshRequired`. Then it feels namespaced and could do `MacaroonRefreshFailed` and similar things if you needed.

I really like using Exceptions to broadcast the state! :+1:

File: modules/publisher/views.py [hunk 6]
```diff
@@ -12,44 +13,50 @@
...
-            else:
-                error_list.append(error)
-
-    user_snaps = []
-    if '16' in account['snaps']:
-        user_snaps = account['snaps']['16']
-
-    flask_user = flask.session['openid']
+    try:
+        account = api.get_account(flask.session)
```
Comment (WillMoggridge) [inline modules/publisher/views.py]: Does the whole function need to be in a try block or can it be limited to a smaller section, maybe this line?

Comment (tbille) [reply modules/publisher/views.py]: For now, yes because this part is not totally refactored yet

File: modules/publisher/api.py (hunk 5 above)
Comment (tbille) [reply modules/publisher/api.py]: `MacaroonRefreshFailed` will be handled in another PR when I arrive on the refactor of this particular file

File: modules/publisher/api.py [hunk 7]
```diff
@@ -127,18 +108,8 @@ def get_account(session):
...
-        ACCOUNT_URL,
-        '/account',
-        '/login',
-    )
-
-    if verified_response is not None:
-        return {
-            'redirect': verified_response
-        }
+    if authentication.verify_headers(response.headers):
```
Comment (WillMoggridge) [inline modules/publisher/api.py]: This function name is pretty confusing but that may be a subject for a different PR
